File: ios/Classes/HttpFactoryController.swift
import Foundation
@_spi(Experimental) import MapboxMaps
import Flutter
import UIKit
class HttpFactoryController: NSObject, HttpFactorySettingsInterface {
    private var pluginVersion: String

    init(withPluginVersion pluginVersion: String) {
        self.pluginVersion = pluginVersion
    }

    func setInterceptor(options: [HttpInterceptorOptions?]) throws {
        HttpServiceFactory.setHttpServiceInterceptorForInterceptor(
            HttpFactoryInterceptor(
                pluginVersion: pluginVersion,
                options: options
            )
        )
    }

    func handleSetInterceptor(methodCall: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let arguments = methodCall.arguments as? [String: Any] else { return }
        guard let options = arguments["options"] as? [HttpInterceptorOptions?] else { return }
        do {
            try setInterceptor(options: options)
        } catch {
            result(FlutterError(code: "HttpFactoryController", message: "Failed to set interceptor", details: nil))
        }
        result(nil)
    }
}

final class HttpFactoryInterceptor: HttpServiceInterceptorInterface {
    private var pluginVersion: String
    private var options: [HttpInterceptorOptions?]

    init(pluginVersion: String, options: [HttpInterceptorOptions?]) {
        self.pluginVersion = pluginVersion
        self.options = options
    }

    func onRequest(for request: HttpRequest,
                   continuation: @escaping HttpServiceInterceptorRequestContinuation) {
        if let oldUserAgent = request.headers["user-agent"] {
            request.headers["user-agent"] = "\(oldUserAgent) FlutterPlugin/\(self.pluginVersion)"
        }

        for option in options {
            guard let option = option else { continue }
            let urlRegex = try! NSRegularExpression(pattern: option.urlRegex)
            let range = NSRange(location: 0, length: request.url.utf16.count)
            if urlRegex.firstMatch(in: request.url, options: [], range: range) != nil {
                let headers = option.headers ?? []
                for header in headers {
                    request.headers[header!.key] = header!.value
                }
            }
        }
        continuation(HttpRequestOrResponse.fromHttpRequest(request));
    }

    func onResponse(for response: HttpResponse, continuation: @escaping HttpServiceInterceptorResponseContinuation) {
        continuation(response);
    }

}

File: android/src/main/kotlin/com/mapbox/maps/mapbox_maps/HttpFactoryController.kt
package com.mapbox.maps.mapbox_maps

import com.mapbox.common.DownloadOptions
import com.mapbox.common.HttpHeaders
import com.mapbox.common.HttpRequest
import com.mapbox.common.HttpRequestOrResponse
import com.mapbox.common.HttpResponse
import com.mapbox.common.HttpServiceFactory
import com.mapbox.common.HttpServiceInterceptorInterface
import com.mapbox.common.HttpServiceInterceptorRequestContinuation
import com.mapbox.common.HttpServiceInterceptorResponseContinuation
import com.mapbox.maps.pigeons.FLTHttpFactorySettings
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

class HttpFactoryController(private val pluginVersion: String) :
  FLTHttpFactorySettings.HttpFactorySettingsInterface {
  override fun setInterceptor(options: MutableList<FLTHttpFactorySettings.HttpInterceptorOptions>) {
    HttpServiceFactory.getInstance().setInterceptor(
      object : HttpServiceInterceptorInterface {
        override fun onRequest(
          request: HttpRequest,
          continuation: HttpServiceInterceptorRequestContinuation
        ) {
          request.headers[HttpHeaders.USER_AGENT] =
            "${request.headers["user-agent"]} Flutter Plugin/$pluginVersion"
          for (option: FLTHttpFactorySettings.HttpInterceptorOptions? in options) {
            if (option == null) {
              continue
            }
            if (request.url.matches(
                Regex(
                  option.urlRegex,
                  RegexOption.IGNORE_CASE
                )
              )
            ) {
              val headers: List<FLTHttpFactorySettings.HttpHeader>? =
                option.headers
              if (headers != null) {
                for (header: FLTHttpFactorySettings.HttpHeader in headers) {
                  request.headers[header.key] = header.value
                }
              }
            }
          }
          continuation.run(HttpRequestOrResponse(request))
        }

        override fun onResponse(
          response: HttpResponse,
          continuation: HttpServiceInterceptorResponseContinuation
        ) {
          continuation.run(response)
        }
      }
    )
  }

  fun handleSetInterceptor(call: MethodCall, result: MethodChannel.Result) {
    val options =
      call.argument<List<FLTHttpFactorySettings.HttpInterceptorOptions>>("options")
    if (options != null) {
      setInterceptor(options.toMutableList())
    }
    result.success(null)
  }
}
